Softmax + LayerNorm + BatchNorm Heuristics

rdspring1 opened the pull request:
- [x] Create Multiple Reduction Schedule Heuristics
- [X] Fuse Point-wise Ops with Softmax via PyTorch JIT
- [X] Fuse Point-wise Ops with Layer Normalization via PyTorch JIT
- [x] Fuse Point-wise Ops with Batch Normalization via PyTorch JIT
- [x] Benchmark performance against eager mode

rdspring1 committed:
Initial Multiple Reduction Scheduler for Softmax and BatchNorm

Fix incorrect recursive error in findCompatibleInputAllocate

rdspring1 committed:
Remove custom schedule from BatchNorm test

rdspring1 committed:
Create custom parser rule for Softmax

Generate launch parameters via multipleReductionHeuristic

rdspring1 committed:
Create custom parser rule for LayerNorm

rdspring1 committed:
Implement softmax_warp_forward heuristics

rdspring1 committed:
Implement Non-Persistent Multiple Reduction Schedule

rdspring1 committed:
Duplicate TensorViews to avoid recompute ComputeAt error in Non-Persistent Kernels

Support reduction along any axis beyond the fastest dimension

rdspring1 committed:
Code Cleanup

rdspring1 committed:
Create custom parser rule for BatchNorm

Infer Reduction Size for IterDomain extent

rdspring1 committed:
Code Style Cleanup

rdspring1 committed:
Streamline findTensorViewsToComputeAtInline

rdspring1 committed:
Added Softmax, Batch_Norm, Layer_Norm Python Unit Tests

rdspring1 committed:
Cache After Transformation Replay

Create LayerNorm cpp test

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+  broadcast_mask[kReductionAxis] = true;
+
+  TensorView* max_val = reductionOp(
+      BinaryOpType::Max, {kReductionAxis}, new Float(FLT_MIN), input);
```
Can you add a max op to arith.cpp?

jjsjann123 commented on torch/csrc/jit/codegen/cuda/kernel_cache.h:
```diff
@@ -168,8 +168,8 @@ class FusionExecutorCache {
 
-  //! cache reduction_tv_ to avoid searching repetitively at runtime
-  TensorView* reduction_tv_ = nullptr;
+  //! cache reduction_tv to avoid searching repetitively at runtime
```
nitpick: `reduction_tv_`

jjsjann123 commented on torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp:
```diff
@@ -137,7 +137,8 @@ class AllocateReuseModifier {
     // Assume the first argument contains the primary variable
     // Follow path along point-wise operations
-    if (first_tv_input != nullptr) {
+    if (first_tv_input != nullptr &&
```
nitpick: maybe update the comment as well since the logic is updated.

rdspring1 replied:
This is a bug fix rather than a change to the core logic.

jjsjann123 commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+              bias = value_map[node->input(2)->unique()]->as<TensorView>();
+            }
+
+            TensorView* running_mean = nullptr;
```
This doesn't look right. We are not updating running_mean/running_var? I don't know we support I/O tensor, so I actually don't know how to do that.

rdspring1 replied:
What is I/O tensor?

jjsjann123 replied:
Basically running_mean should be updated as:
`running_mean = running_mean * ( 1 - momentum ) + mean * momentum`

FYI, running stats are only used during inference, our test ground should cover that as well. We should do a few iterations of training followed by a few iterations of inference and compare with reference.

rdspring1 replied:
I think we'd need global memory aliasing to handle the I/O tensor. Currently we have generate new tensor for the new_running_mean. I only partially implemented batch norm in this PR.

```
new_running_mean = running_mean * ( 1 - momentum ) + mean * momentum
running_mean = new_running_mean
```

jjsjann123 commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+            }
+            value_map.emplace(node->output()->unique(), output);
+
+            // Merge reduction axes together
```
Looks like that we are mixing computation with scheduling. I would try to avoid this if possible.

rdspring1 replied:
For scheduling, we need to merge all non-reduction and reduction axes together and to place the merged reduction axis in the -1 position.

1) Softmax - Single Axis Reduction
[N,  C,  H,  W] => (mark reduction axis) [N, rC,  H,  W] => (merge then reorder) [N, H * W, rC]

2) LayerNorm - Contiguous, Multiple Reduction Axes
[N,  C,  H,  W] => (mark reduction axes) [N, C, rH, rW] => (merge) [N*C, rH * rW]

3) BatchNorm - Non-Contiguous, Multiple Reduction Axes
[N,  C,  H,  W] => (mark reduction axes) [rN, C, rH, rW] => (reorder then merge) [C, rN * rH * rW]

If you wanted to handle this schedule.cpp, you would need to infer which case you have before transforming the TensorViews. This information is already given in parser.cpp.

jjsjann123 replied:
Totally agree that this is a convenient place to do that. But putting transformation in parser means that scheduler needs to be made aware of pre-schedule transformations.
I can live with us taking the short cut, but maybe we should re-name multiple reduction scheduler more specifically to normalization scheduler and restricting its application to only those operations? Things that would need to be done is to define normalization operations (expand this: https://github.com/csarofeen/pytorch/blob/52581a0bf13f1a92e8346e6fbb08c7f44fcb9d55/torch/csrc/jit/codegen/cuda/parser.cpp#L163-L166 to identify normalization, as well as updating `graph_fuser.cpp`). Alternatively we can put this into our fusion partition (if we are going that route instead). This should be done in a separate PR tho.

Side note, complications from this:
Permutation on stride in integration doesn't work with this, because we don't have reduction axes for normalization layers. This is a problem we need to solve. An open question is, should we wait until we have permutation in codegen? [I'll open an issue tagging NHWC]

jjsjann123 commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+              }
+            }
+          });
+    }
```
We should mark normalization as (reduction) operations: https://github.com/csarofeen/pytorch/blob/52581a0bf13f1a92e8346e6fbb08c7f44fcb9d55/torch/csrc/jit/codegen/cuda/parser.cpp#L153-L157

jjsjann123 reviewed (changes requested):
I'm quite curious on how general we think the multiple reduction scheduling is. Does it actually supports arbitrary computational graph with reduction in it?
My impression is that this is something specifically tailored for normalization, we might want to have the name to reflect that.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+      {bcast_max, x_max_sub, exp, bcast_sum, output});
+
+  auto reduction_params =
+      getMultipleReductionHeuristics(&fusion, {t0}, reduction_tensors);
```
Why did you modify the fusion before sending it to heuristics/scheduling? It doesn't look like this is done during the fusion pass.

rdspring1 replied:
I need to place the reduction axis in the inner-most dimension because of `rfactor` and `computeAt`. I handle this rearrangement at a per-op level, not in the scheduler.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+    const int axes = input->nDims() - 1 - idx;
+    reduction_axes[idx] = axes;
+    broadcast_mask[axes] = true;
+  }
```
Need to check if we can/if this PR always assumes inner most axes normalization.

rdspring1 replied:
Isn't this a property of layer_norm?

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+  // Failure Case - Multiple inputs to same expression
+  // auto x_mean_sub_pow = mul(x_mean_sub, x_mean_sub);
+
+  auto x_mean_sub_pow = binaryOp(BinaryOpType::Pow, x_mean_sub, new Float(2));
```
Don't use Pow 2, use mul. Pow has had perf issues in the past in normalization.

rdspring1 replied:
`mul` was giving me trouble with `duplicate`, and the fix was cumbersome. It also messes with the invariant of `duplicate` - the number of uses for a TensorView after `duplicate` is 1.

csarofeen replied:
File an issue on this please.

rdspring1 replied:
Created https://github.com/csarofeen/pytorch/issues/516

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+  // auto new_running_mean = add(rmean_hat, current_mean_hat);
+
+  auto x_mean_sub = sub(input, x_mean);
+  auto x_mean_sub_pow = binaryOp(BinaryOpType::Pow, x_mean_sub, new Float(2));
```
Use mul

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+      broadcast_mask[axis] = true;
+      num_features = (axis == 0)
+          ? input->domain()->domain()[0]->extent()
+          : mul(num_features, input->domain()->domain()[axis]->extent());
```
Can this approach be used for layer norm as well?

rdspring1 replied:
For layer norm, we are given a list of ints, which represents the number of features for normalization. I can calculate that in parser.cpp. On second thought, this would require generating a new kernel for a new size, so this dynamic approach is better. :man_shrugging:

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+  at::Tensor tmean = at::zeros({input_shape[1]}, options);
+  at::Tensor tvar = at::ones({input_shape[1]}, options);
+
+  // Check reduction axis is same for all reductions
```
It would be interesting to see if we can use iterdomain mapping @naoyam worked on to do this particular check.

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -649,6 +650,127 @@ def test_reduction(self):
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_batch_norm(self):
+        output_elements = 10000
```
can we decrease this to 5k? Each input tensor will be ~1.3GB and I don't recall how much memory upstream CI has with their GPUs.

jjsjann123 replied:
If we hit issue with OOM we can guard it with:
`@unittest.skipIf(not TEST_LARGE_TENSOR, "not enough memory")`

Upstream usually keep things small for python test. It makes sense for us as well since end-2-end tests focuses on integration code. If we worry about codegen correctness let's do that in cpp tests.

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -649,6 +650,127 @@ def test_reduction(self):
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_softmax(self):
+        output_size = 10000
```
Same here.

csarofeen commented on torch/csrc/jit/codegen/cuda/compute_at.cpp:
```diff
@@ -39,8 +39,8 @@ void ComputeAtData::setPassPosition(unsigned int pos) {
-        "Error during computeAt. ComputeAt pass wanted to set position of ",
-        tv_ref_,
+        "Error during computeAt. ComputeAt pass wanted to set position of TensorView: ",
+        tv_ref_->name(),
```
Are the tensor views not displaying properly from these errors? CC @naoyam

rdspring1 replied:
It displays the pointer address, not the name of the TensorView.

naoyam replied:
It sometimes happens to me as well. I tried to understand when it happens, but gave up. This is part of the reason @tlemo added the explicit stringfication function for the kernel IR (`kir::toString`).

csarofeen replied:
Do we have an issue open for this?

naoyam replied:
No, I don't think so. But I created a test before. https://github.com/csarofeen/pytorch/blob/c10-str-test/test/cpp/jit/test_gpu.cpp#L7090-L7097

And changed `TensorView::TensorView` as:
https://github.com/csarofeen/pytorch/blob/c10-str-test/torch/csrc/jit/codegen/cuda/tensor_view.cpp#L29-L30

The result was:
```
[ RUN      ] NVFuserTest.Stringify
In ctor: 0x563e772dedb0
In ctor (const*): T0[ iS0{i1} ]
c10::str(TensorView*): T0[ iS0{i1} ]
c10::str(const TensorView*): T0[ iS0{i1} ]
```

naoyam replied:
Recorded as an issue (#514)

csarofeen commented on torch/csrc/jit/codegen/cuda/fusion.cpp:
```diff
@@ -279,12 +279,13 @@ void Fusion::validateInputs() {
+    if (!input->isConstScalar()) {
       TORCH_CHECK(
-          hasInput(input),
+          hasInput(input) || inFusion(input),
```
inputsOf should return all roots of out. Was there a case where it was returning an intermediate?

rdspring1 replied:
The dynamic size for batch norm is not an input to the fusion or an intermediate tensor. It complains about this in `validateInputs`

csarofeen replied:
Can you file an issue for this?

rdspring1 replied:
Created https://github.com/csarofeen/pytorch/issues/517

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+  FusionGuard fg(fusion());
+
+  TORCH_INTERNAL_ASSERT(
+      !hasComputeAt(), "Duplicate Tensor after defining ComputeAt structure");
```
Error message isn't clear, can you rephrase to "Cannot duplicate tensor after defining ComputeAt structure."

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+  auto usages = fusion()->unordered_uses(this);
+  TORCH_INTERNAL_ASSERT(
+      usages.size() > 1,
+      "Error duplicating TensorView: multiple usages is required");
```
Rephrase please: "Cannot duplicate tensor view that is only used once."

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+      usages.size() > 1,
+      "Error duplicating TensorView: multiple usages is required");
+
+  // Warning: error may occur if the same TensorView
```
We should explicitly check for duplicate uses in exprs and handle them correctly i.e. don't duplicate for uses within the same expr.

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+  std::vector<TensorView*> duplicates;
+  Expr* origin_expr = fusion()->origin(this);
+  size_t count = 0;
+  for (auto expr : usages) {
```
Should we assume to duplicate for every usage? Or should the interface specify which expr to create a duplicate for?

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+std::vector<TensorView*> TensorView::duplicate() {
```
If this is an input or output we should throw an error.

csarofeen replied:
This interface seems strange for "recomputing" a tensor as it doesn't recompute the entire graph producing a TensorView, only a single expression.

csarofeen replied:
Should duplicated tensors be replayed as the tensor they duplicated? What about logic like computeAt of the duplicate?

rdspring1 replied:
Do you want a single TensorView op to manipulate the entire graph?

rdspring1 replied:
I call `TransformReplay::fullSelfReplay` to transform the duplicated tensor.

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -595,11 +632,17 @@ TensorView* TensorView::cache_after() {
     for (auto expr : fusion()->unordered_uses(consumer)) {
       for (TensorView* output :
            ir_utils::filterByType<TensorView>(expr->outputs())) {
         if (output->hasComputeAt()) {
```
I'm really confused by the logic here. Why would we computeAt a tensor we're caching before if it's outputs have a compute at but the tensor we're caching didn't?

rdspring1 replied:
When caching an input (the most common use case), it doesn't have a computeAt even if you call `inputs->computeAt(outputs, dim)`.

csarofeen replied:
Then please guard this computeAt setting with if it's an input.

rdspring1 replied:
:+1:

csarofeen commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -595,11 +632,17 @@ TensorView* TensorView::cache_after() {
         if (output->hasComputeAt()) {
+          if (!cache_replayed) {
+            // Completely transform consumer according to output
+            TransformReplay::replayPasC(consumer, output, -1);
```
Why would we do this?

rdspring1 replied:
We have transformations to merge non-reduction and reduction axes together. Those transformations need to be replayed for all dimensions, otherwise I need to manually replay them.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+  // Check reduction axis is same for all reductions
+  // Generate Launch Parameters
+  auto reduction_params =
+      getMultipleReductionHeuristics(&fusion, {t0}, reduction_tensors);
```
Why did you modify the fusion before sending it to heuristics/scheduling? It doesn't look like this is done during the fusion pass.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -6610,6 +6610,328 @@ TEST(NVFuserTest, FusionSmemDynamicPersistentSoftmax2D_CUDA) {
+
+  // Check reduction axis is same for all reductions
+  // Generate Launch Parameters
+  auto reduction_params = getMultipleReductionHeuristics(
```
Why did you modify the fusion before sending it to heuristics/scheduling? It doesn't look like this is done during the fusion pass.

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -649,6 +650,127 @@ def test_reduction(self):
                         perm1 = range(len(x))
                         self._reduction_helper(x, axes, torch.float32, "cuda", perm0, perm1, keepdim)
 
+    def _layer_norm_helper(self, shape, norm_shape, dtype, device):
```
@jjsjann123 did you review this file?

csarofeen commented on torch/csrc/jit/codegen/cuda/shape_inference.cpp:
```diff
@@ -147,6 +147,29 @@ class NaiveTypePropagator {
         node->output()->setType(promoted_type);
         break;
       }
+      case aten::batch_norm: {
```
@jjsjann123 did you review this file?

rdspring1 commented:
The schedule is tailored from a performance perspective for softmax. The main assumption is that the reduction axes are the same for every reduction. The secondary performance assumption is that the number of elements in the reduction axes is the same. I think this schedule would work with [torch.vmap](https://pytorch.org/docs/master/generated/torch.vmap.html).

naoyam commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -595,11 +632,17 @@ TensorView* TensorView::cache_after() {
       for (TensorView* output :
            ir_utils::filterByType<TensorView>(expr->outputs())) {
         if (output->hasComputeAt()) {
+          if (!cache_replayed) {
+            // Completely transform consumer according to output
+            TransformReplay::replayPasC(consumer, output, -1);
+            cache_replayed = true;
+          }
```
Can you comment why you're doing this?

rdspring1 replied:
When scheduling softmax and normalization ops, we need to merge all non-reduction and reduction axes together and to place the merged reduction axis in the -1 position.

If we have a point-wise operation before normalization, we don't apply those necessary transformations to the cache_after tensor. We use `TransformReplay::replayPasC` to apply all the transformations from the input's consumers.

naoyam replied:
Sorry, I don't understand why that needs to be done as part of `cache_after`. I'm not 100% sure if we really need this part of `cache_after` (see #483). 

Is this merely for convenience? Or is it required for correctness?

rdspring1 replied:
Do you not want to support calling cache-after and cache-before after creating computeAt structure? If so, why not?

Without transform-replay the user has to go back and manually transform the cache?

If the transformation is already known, why not have the cache function transform the Tensorview?

Also, the normalization transforms are handled in parser.cpp, so they are not known in schedule.cpp.

naoyam replied:
> Do you not want to support calling cache-after and cache-before after creating computeAt structure? If so, why not?
> 

I haven't thought about it. Why are you asking this?

rdspring1 replied:
Then what are you referring to with issue [#483](https://github.com/csarofeen/pytorch/issues/483)? 

It is only an issue if apply cache-after after setting computeAt.

naoyam replied:
I suspect that #483 was confusing because the line numbers need to be updated. It was originally referring to the `else` part of `if (hasComputeAt())`. https://github.com/csarofeen/pytorch/blob/20_10_20_devel/torch/csrc/jit/codegen/cuda/tensor_view.cpp#L598-L610

naoyam commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+std::vector<TensorView*> TensorView::duplicate() {
```
Can you add tests just for this?

naoyam commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+std::vector<TensorView*> TensorView::duplicate() {
+  FusionGuard fg(fusion());
+
+  TORCH_INTERNAL_ASSERT(
```
Shouldn't this be a user error rather than an internal error as it's a contract of this function?

naoyam commented on torch/csrc/jit/codegen/cuda/tensor_view.cpp:
```diff
@@ -391,6 +391,43 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
+      !hasComputeAt(), "Duplicate Tensor after defining ComputeAt structure");
+
+  auto usages = fusion()->unordered_uses(this);
+  TORCH_INTERNAL_ASSERT(
```
ditto

naoyam commented on torch/csrc/jit/codegen/cuda/ir_interface_nodes.h:
```diff
@@ -301,6 +301,11 @@ class TORCH_CUDA_API TensorView : public Val {
+  // Create a new TensorView for all usages of this TensorView.
+  // A common use case is to handle the recompute ComputeAt exception that
+  // occurs when inlining a TensorView used multiple times in a fusion.
+  std::vector<TensorView*> duplicate();
```
Please add comments on the constraints about `this` (e.g., having no computeAt view). 

Also, I think it's important to mention here that this duplicates not only `this` but also its origin expression.

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+
+            auto x_mean_sub = sub(input, x_mean);
+            auto x_mean_sub_pow =
+                binaryOp(BinaryOpType::Pow, x_mean_sub, new Float(2));
```
change to mul please

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+
+            /*
+            auto current_mean_hat = mul(x_mean, new Float(kMomentum));
+            auto rmean_bcast = broadcast(running_mean, broadcast_mask);
```
I know this isn't enabled yet, is the broadcast to actually broadcast or you just need the extra dimension?

rdspring1 replied:
I'm confused. What is the difference?

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+
+            // Merge reduction axes together
+            // Reorder axes such that reduction axis is at inner-most
+            // dimension [Outer, Reduction]
```
I think as a general rule this should be done in the scheduler.

rdspring1 replied:
This transformation is different for softmax, batch_norm, and layer_norm. This information is lost in the scheduler and needs to be inferred again.

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+            for (auto size : norm_shape->vec()) {
+              numel *= size;
+            }
+            Int* N = new Int(numel);
```
Is the reason you use this approach versus the previous with `IterDomain->extent()` because layer norm can only normalize to a fixed shape in the front end?

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+                   ++idx) {
+                tv->merge(0, 1);
+              }
+            }
```
Again, I think this should be done in the scheduling, not on the original fusion.

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,350 @@ class IrParser {
+                  tv->reorder({{0, 1}, {1, 0}});
+                }
+              }
+            }
```
Same, should be moved to scheduler.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -142,12 +142,134 @@ constexpr int lastPow2(int n) {
+    int64_t outer_dim_size,
+    int64_t inner_dim_size,
+    bool fastest_dim_reduction) {
+  if (fastest_dim_reduction) {
```
Does softmax not work on a 1D input or running across all dimensions of an N-D tensor?

rdspring1 replied:
:man_shrugging: I took the assumption from the reduction heuristics.

csarofeen replied:
Can you open an issue. We're trying to sort that out with the reduction heuristics now (https://github.com/csarofeen/pytorch/pull/496) and will want to do the same here. I assume this wouldn't be a problem with Batch Norm or Layer Norm as they always assume there's some axes we aren't normalizing across. Is that a correct assumption?

rdspring1 replied:
I enabled softmax to reduce across all dimensions. It is the case when the outer and inner is zero. Batch Norm and Layer Norm should always have an outer or inner dim size.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -142,12 +142,134 @@ constexpr int lastPow2(int n) {
+  if (fastest_dim_reduction) {
+    TORCH_INTERNAL_ASSERT(reduction_dim_size > 0 && outer_dim_size > 0);
+  } else {
+    TORCH_INTERNAL_ASSERT(
```
Similar question here as the previous. We do support full reductions to a scalar value.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -142,12 +142,134 @@ constexpr int lastPow2(int n) {
+
+  // Is fastest dimension a reduction dimension?
+  if (rparams.fastest_dim) {
+    // TODO: multiple batches per warp causes layer-norm errors
```
Do you suspect there's a codegen bug causing this or something with the schedule?

rdspring1 replied:
I think it is a codegen, launch parameters, schedule bug, because it only fails on certain offsets when reducing on the fastest dimension and it occurs when I split the outer dimension.

e.g. The schedule and kernel is the same for all these examples
1. [20, 100, 35, 67] pass
2. [20, 100, 35, 68] pass
3. [20, 100, 35, 69] fail
4. [20, 100, 35, 70] fail
4. [20, 100, 35, 71] pass

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -142,12 +142,134 @@ constexpr int lastPow2(int n) {
+    bdimx = 1;
+    if (bdimy <= kBlockThresholdNotFastestDim &&
+        reduction_dim_size >= kBlockThresholdNotFastestDim) {
+      while (bdimy * bdimx <= kMaxThreadsPerCTA &&
```
Instead of dividing later can't you just change the while  to `bdimy * bdimx * 2 <= kMaxThreadsPerCTA && bdimx * 2 <= reduction_dim_size`. Uncertain if it's possible but if the while never gets triggered and `bdimx = 1` then `bdimx /=2` will be 0.

rdspring1 replied:
I'll take a look. I just copy & pasted from [SoftMax_getBlockSize](https://github.com/pytorch/pytorch/blob/3ffd2af8cd7b52c95177576b1246de546488af02/aten/src/ATen/native/cuda/SoftMax.cu#L144).

rdspring1 replied:
The while loop should always trigger.
For the first iteration:
```
bdimy <= 64 (kBlockThresholdNotFastestDim) <= 1024 (kMaxThreadsPerCTA)
1 <= bdimx < 64 (kBlockThresholdNotFastestDim) <= reduction_dim_size
```
I added a guarantee `bdimx = std::max(bdimx, 1)`

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -142,12 +142,134 @@ constexpr int lastPow2(int n) {
+    // First, tile blocks over the y-axis
+    gdimy = std::min(ceilDiv(inner_dim_size, bdimy), kMaxActiveBlocks);
+    // Then, fill the x-axis with remaining blocks
+    gdimx = std::min(ceilDiv(kMaxActiveBlocks, gdimy), outer_dim_size);
```
If you're trying to get a wave of blocks shouldn't you change the grid y and x dimension. Pretty confident we iterate over x blocks before y.

rdspring1 replied:
I also took this from [SpatialSoftmax](https://github.com/pytorch/pytorch/blob/3ffd2af8cd7b52c95177576b1246de546488af02/aten/src/ATen/native/cuda/SoftMax.cu#L107-L142).

csarofeen replied:
yeah that seems wrong, can leave it for now but pretty sure the block ordering should be reversed.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -288,6 +412,85 @@ ReductionParams reductionHeuristic(
+  }
+
+  // Check Reduction Invariants
+  for (auto tv : reduction_tv) {
```
Should you also check that `reduction_tv` is not empty?

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -288,6 +412,85 @@ ReductionParams reductionHeuristic(
+  // Check that the dimensions of the reductions are equal
+  TORCH_INTERNAL_ASSERT(
+      fastest_dim_reduction.size() > 1,
+      "A single reduction tv was detected. Use getReductionHeuristics.");
```
Can't this check go before you gather all the sizes, in your first loop on reduction_tv that does more basic sanity checking?

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+
+//! Find all TensorViews that require duplication to avoid recompute
+//! computeAt error when applying inline ComputeAt
+std::vector<TensorView*> findTensorViewsToDuplicate(
```
@naoyam curious about your thoughts on this. Maybe you should file an issue to re-evaluate and maybe refactor this to a less conservative approach. I also don't recall if I pinged you on the `TensorView::duplicate` logic, but would appreciate it if you would review that too.

naoyam replied:
We already have a related issue (#139). `ComputeAtRootDomainMap` can tell if recomputation is necessary, so I think it'd be possible to extend computeAt to replicate tensors and expressions automatically.

naoyam replied:
https://github.com/csarofeen/pytorch/pull/440#issuecomment-725611237

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+        if (!domain[axis]->isBroadcast() && !domain[axis]->isReduction()) {
+          constant_allocation &= domain[axis]->isConstScalar();
+        }
+      }
```
This only works if the tensor view memory type is local. Should you guard on that in this function?

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+}
+
+//! Place all cache TensorViews in Shared Memory
+//! All point-wise TensorViews inherit shared memory from their parents
```
Why is this the case? Should we not only care to cache the input values?

rdspring1 replied:
This function was when I didn't support computeAt inline, so all those TensorViews with non-static allocation sizes need to be in shared memory.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.h:
```diff
@@ -51,7 +60,7 @@ class ReductionParamsHash {
         static_cast<size_t>(rp.cross_block) << (bits - 2) |
         static_cast<size_t>(rp.cross_grid) << (bits - 3) |
-        static_cast<size_t>(rp.mul_reds_per_blk) << (bits - 4);
+        static_cast<size_t>(rp.multiple_reds_per_blk) << (bits - 4);
```
Missing batches per block, num_warps, and persistent? Should we add them?

rdspring1 replied:
@jjsjann123 Should I add `batches_per_block` and `persistent` to `ReductionParamsHash`?

csarofeen replied:
As soon as we start using them we should

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+      // 2) Apply heuristics to each reduction
+      std::vector<TensorView*> rfactor_tv;
+      for (auto tv : reduction_tv) {
+        // Reduction Splits
```
These are splits on non-reduction IterDomains but reduction tensor views correct? Could a comment syntax like Kevin used with single reduction heuristics be used here? It may help make the logic a little more clear.

csarofeen replied:
Looks like this will split both iter domains and reduction domains

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+      for (auto tv : reduction_tv) {
+        // Reduction Splits
+        // [output-Leftover, Batches-Per-Block, Warp, Loop-Unroll]
+        if (rparams.batches_per_block > 1) {
```
Should we support split's of 1 that we then simplify from the loop nest generation? Would that make this logic simpler?

rdspring1 replied:
Wouldn't changing the loop nest generation affect the computeAt position?

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+          tv->split(0, rparams.batches_per_block);
+        }
+        if (rparams.num_warps > 1) {
+          tv->split(1, rparams.num_warps);
```
The position doesn't change if there was a split from batches_per_block?

rdspring1 replied:
The splits - `batches_per_block` and `num_warps` go together.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+      }
+
+      // 4) ComputeAt Structure
+      const int kComputeAtAxis = 1;
```
I think this can be 2 if `rparams.batches_per_block > 1` there was an smem bug before that probably prevented this from working, but should be fixed now.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+          other_tv.push_back(input->cache_after());
+        }
+      } else {
+        auto compute_inline_tv =
```
This is pretty cool logic.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+        tv->axis(0)->parallelize(ParallelType::BIDx);
+        if (rparams.num_warps > 1) {
+          tv->axis(2)->parallelize(ParallelType::TIDy);
+        }
```
I assume num_warps and batches_per_block are either both on or both off?

csarofeen replied:
I guess for now you're not really using batches_per_block > 1

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+        tv->axis(-2)->parallelize(ParallelType::TIDx);
+      }
+    }
+  } else {
```
This else is really far down, can you just comment this is the non-fastest-dim logic?

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+      if (rparams.lparams.bdimx() > 1) {
+        tv->split(
+            reduction_axis, NamedScalar::getParallelDim(ParallelType::TIDx));
+      }
```
Isn't there a TV between the reduction and broadcast that will be missing a dimension? I'm missing the logic that handles that.

rdspring1 replied:
Hmm, I don't understand. What TV are you thinking of?

csarofeen commented:
> The schedule is tailored from a performance perspective for softmax. The main assumption is that the reduction axes are the same for every reduction. The secondary performance assumption is that the number of elements in the reduction axes is the same. I think this schedule would work with [torch.vmap](https://pytorch.org/docs/master/generated/torch.vmap.html).

I think these are things we can now infer from our root domain mapping logic @naoyam to think about how we might do this.

rdspring1 committed:
Fix Minor Comments

rdspring1 committed:
Dynamic num_features size in layer_norm

rdspring1 committed:
Organize Axes in schedule instead of parser definition

Merge Reduction Axes [Inner => Outer]

rdspring1 committed:
Merge Reduction Axes [Outer => Inner]

rdspring1 committed:
Create max + min op

rdspring1 committed:
Support fastest-dim reduction to scalar value

csarofeen commented on benchmarks/myscripts.txt:
```diff
@@ -0,0 +1,5 @@
+python -m tensorexpr layernorm --device gpu --mode fwd --cuda_fuser=nvf
```
Did you intend to keep this file?

rdspring1 replied:
I removed the file.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler.cpp:
```diff
@@ -670,6 +885,461 @@ void scheduleReduction(
+      int inner_axis = reduction_axes[idx + 1] - idx;
+      tv->merge(kOuterAxis, inner_axis);
+    }
+  }
```
Are you content with the reduction axis moving to the left when getting merged, or should we support a case where we actually move it to the right? Don't think it would be too hard to modify merge so it moves to the right if you think that would be better.
Just wondering about the case where we have:
[I, R, I, R]
We would flag the reduction axis as being inner most, but we would end up with an iter axis to the right of the reduction. Would the scheduling logic get confused about this?

rdspring1 replied:
Scheduling configuration is determined by the root domain. There is also a check to move the merged reduction axis to the inner-most position later in the function.

naoyam commented:
`TensorView::duplicate` looks fine to me. I'm just unsure whether it should be exposed to users. Ideally, it should be just automatically taken care during computeAt.

jjsjann123 commented on test/test_jit_cuda_fuser.py:
```diff
@@ -649,6 +650,127 @@ def test_reduction(self):
+        running_mean = torch.randn(shape[1], dtype=dtype, device=device)
+        running_var = torch.randn(shape[1], dtype=dtype, device=device)
+        t_jit = torch.jit.script(t)
+        jit_o = t_jit(x, y, running_mean, running_var)
```
nitpick: I know we don't have `running_mean` / `running_var` updated yet, but we could still put the right code here. Each iteration should have their own clone of `running_mean` / `running_var` and put a comment that we should validate those results once we have the implementation.

rdspring1 replied:
:+1:

jjsjann123 commented on test/test_jit_cuda_fuser.py:
```diff
@@ -649,6 +650,127 @@ def test_reduction(self):
+    @unittest.skipIf(not RUN_CUDA, "requires CUDA")
+    @unittest.skipIf(GRAPH_EXECUTOR != ProfilingMode.PROFILING,
+                     "Requires fusion optimization pass to be effective")
+    def test_layer_norm(self):
```
Can we add a simple test for fp16 as well?
just to cover scheduling with an added cast op after the second reduction.

rdspring1 replied:
:+1:

rdspring1 committed:
Added FP16 Python Tests

Treat FP16 cast as cache_after in persistent kernels

Add Batch Norm Running Mean + Var Checks

Enable cache_after computeAt lookahead only for fusion inputs

jjsjann123 commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -451,6 +451,232 @@ class IrParser {
+            auto var = div(var_sum_bcast, num_features);
+
+            /*
+            auto current_var_hat = mul(var, new Float(kMomentum));
```
Somehow my comments are not posted.......

I think this is intended for running_var update? Should use unbiased variance for running stats.
```
auto unbiased_var = div(var_sum_bcast, sub(num_features, new Int(1)))
auto current_var_hat = mul(unbiased_var, new Float(kMomentum));
```

rdspring1 replied:
:+1:

rdspring1 committed:
Improve Multiple Reduction Split Comments

rdspring1 committed:
Unbiased variance for batch norm statistics

rdspring1 committed:
Initial Fastest Dim Vectorized Schedule

rdspring1 committed:
Tweak Block Size to support FP16 Batch Norm

rdspring1 committed:
Duplicate TensorViews with Non-Constant Allocations using ComputeAt structure

csarofeen reviewed (approved):
Looks good, thanks for the careful revisions after review.

jjsjann123 reviewed (approved):
Requested changes have been made.
Note to myself that we need to do a follow up on partition update.

rdspring1 committed:
Update ReductionParamsHash with latest parameters

rdspring1 committed:
Fix TensorView::Duplicate to replace all the same arguments with new duplicate
